docs(main): update data fetch doc (#3737)
web-infra-dev/modern.js: packages/document/main-doc/docs/en/guides/basic-features/data-fetch.mdx

---
title: Data Fetching
sidebar_position: 3
---
# Data Fetching

Modern.js provides out-of-the-box data fetching capabilities, allowing developers to develop in an isomorphic way in both client-side and server-side code.

It should be noted that these APIs do not help applications initiate requests, but rather help developers better manage data and improve project performance.

## Data Loader (Recommended)

Modern.js recommends using [conventional routing](/guides/basic-features/routes) for routing management. Through Modern.js's [conventional (nested) routing](/guides/basic-features/routes#conventional-routing), each routing component (`layout.ts` or `page.ts`) can have a same-named `loader` file. The `loader` file needs to export a function that will be executed before the component is rendered to provide data for the routing component.

:::info

> Modern.js v1 supports getting data through [useLoader](/guides/basic-features/data-fetch.html#useloader-(old-version)), which is no longer the recommended usage. We do not recommend mixing the two except during the migration process.

Reviewer: ```suggestion
Modern.js v1 supports fetching data via [useLoader](/guides/basic-features/data-fetch.html#useloader-(old-version)), which is no longer the recommended usage. We do not recommend mixing the two except during the migration process.
```

};
```

4. When running on the server, the `loader` function will be packaged into a unified bundle, so we do not recommend using `__filename` and `__dirname` in server-side code.

### FAQ

1. Relationship between `loader` and BFF functions

In CSR projects, `loader` is executed on the client side, and the BFF function can be called directly in the `loader` to make interface requests.

In SSR projects, each `loader` is also a server-side interface. We recommend using the `loader` instead of the BFF function with an http method of `get` as the interface layer to avoid an extra layer of forwarding and execution.

## useLoader (old version)

> **`useLoader`** is an API in the old version of Modern.js. This API is a React Hook designed specifically for SSR applications, allowing developers to get data in components in isomorphic development.

Reviewer: ```suggestion
**`useLoader`** is a legacy API in Modern.js v1. This API is a React Hook designed specifically for SSR applications, allowing developers to fetch data in components in isomorphic development.
```

};
```

After running the above code, when you access the page, you can see that logs are output to the terminal, but not printed in the browser console.

This is because Modern.js collects the data returned by `useLoader` during server-side rendering and injects it into the corresponding HTML. If the SSR rendering is successful, you can see the following code snippet in the HTML:

```html
<script>
  window._SSR_DATA = {};
</script>
```

> In this global variable, each piece of data is recorded, and during the browser-side rendering process, this data is used first. If the data does not exist, the `useLoader` function will be executed again.

Reviewer: ```suggestion
This global variable is used to record data, and during the browser-side rendering process, this data is used first. If the data does not exist, the `useLoader` function will be executed again.
```
